File: 2021/day01/solution.lua
function run(file)
    local increase, previous = 0, nil
    for line in io.lines(file) do
        local n = tonumber(line)
        if previous and n > previous then
            increase = increase + 1
        end
        previous = n 
    end
    return increase
end

function sum(window)
    local total = 0
    for _, value in ipairs(window) do
        total = total + value
    end
    return total
end

function run_part2(file)
    local increase, previous = 0, nil
    local f = io.open(file)

    local window = {
        f:read("n"),
        f:read("n"),
        f:read("n")
    }

    repeat
        local last = f:read("n")

        local next_window = {
            window[2],
            window[3],
            last
        }

        if sum(next_window) > sum(window) then
            increase = increase + 1
        end

        table.move(next_window, 1, #next_window, 1, window)
    until not last
    
    return increase
end

assert(run("sample.txt") == 7)
assert(run_part2("sample.txt") == 5)

File: 2021/day01/solution.py
def run(file):
    with open(file) as f:
        increase = 0
        previous = None
        for line in f:
            current = int(line)
            if previous and current > previous:
                increase += 1
            previous = current
        return increase


def run_part2(file):
    with open(file) as f:
        lines = f.read().splitlines()
        start, end = 0, 3
        increase = 0
        previous = None

        while True:
            line = lines[start:end]
            if len(line) < 3:
                break
            current = sum(map(int, line))
            if previous and current > previous:
                increase += 1
            previous = current
            start += 1
            end += 1

        last = sum(map(int, lines[-3:]))

        if last > previous:
            increase += 1

        return increase


if __name__ == '__main__':
    assert run("sample.txt") == 7
    assert run_part2("sample.txt") == 5
